hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/TestS3AAWSCredentialsProvider.java

```diff
@@ -480,4 +488,149 @@ public void refresh() {
+        provider.isInitialized());
+    assertFalse(
+        "Provider not initialized. hasCredentials should be false",
+        provider.hasCredentials());
+    if (provider.getInitializationException() != null) {
+      throw new AssertionError(
+          "Provider not initialized. getInitializationException should return null",
+          provider.getInitializationException());
+    }
+
+    for (int i = 0; i < threads; i++) {
+      results.add(pool.submit(() -> list.getCredentials()));
+    }
+
+    for (Future result : results) {
+      AWSCredentials credentials = (AWSCredentials) result.get();
+      assertEquals(credentials.getAWSAccessKeyId(), "expectedAccessKey");
+      assertEquals(credentials.getAWSSecretKey(), "expectedSecret");
+    }
+
+    pool.awaitTermination(10, TimeUnit.SECONDS);
+    pool.shutdown();
+
+    assertTrue(
+        "Provider initialized without errors. isInitialized should be true",
+         provider.isInitialized());
+    assertTrue(
+        "Provider initialized without errors. hasCredentials should be true",
+        provider.hasCredentials());
+    if (provider.getInitializationException() != null) {
+      throw new AssertionError(
+          "Provider initialized without errors. getInitializationException should return null",
+          provider.getInitializationException());
+    }
+  }
+
+  /**
+   * Credential provider with error.
+   */
+  private static class ErrorProvider extends AbstractSessionCredentialsProvider {
+
+    public ErrorProvider(@Nullable URI uri, Configuration conf) {
+      super(uri, conf);
+    }
+
+    @Override
+    protected AWSCredentials createCredentials(Configuration config) throws IOException {
+      throw new IOException("expected error");
+    }
+  }
+
+  @Test
+  public void testConcurrentAuthenticationError() throws Throwable {
+    Configuration conf = createProviderConfiguration(ErrorProvider.class.getName());
+    Path testFile = getCSVTestPath(conf);
+
+    int threads = 10;
+
+    AWSCredentialProviderList list = createAWSCredentialProviderSet(testFile.toUri(), conf);
+    ErrorProvider provider = (ErrorProvider) list.getProviders().get(0);
+
+    ExecutorService pool = Executors.newFixedThreadPool(threads);
+
+    List<Future> results = new ArrayList<>(threads);
+
+    assertFalse("Provider not initialized. isInitialized should be false",
+        provider.isInitialized());
+    assertFalse("Provider not initialized. hasCredentials should be false",
+        provider.hasCredentials());
+    if (provider.getInitializationException() != null) {
+      throw new AssertionError(
+          "Provider not initialized. getInitializationException should return null",
+          provider.getInitializationException());
+    }
+
+    for (int i = 0; i < threads; i++) {
+      results.add(pool.submit(() -> list.getCredentials()));
+    }
+
+    for (Future result : results) {
+      intercept(java.util.concurrent.ExecutionException.class,
```

Comment:
actually, i see we have an `interceptFuture` method which lets you specify the class of the inner exception....